### PR TITLE
ERR: add stacklevel to to_dict() UserWarning (#16927)

### LINKED ISSUE #16927: ERR: missing stacklevel in to_dict UserWarning
#### Code Sample, a copy-pastable example if possible

```python
In [4]: df = pd.DataFrame([[1,2,3]], columns=['a', 'a', 'b'])

In [5]: df
Out[5]:
   a  a  b
0  1  2  3

In [6]: df.to_dict()
C:\<snip>\pandas\core\frame.py:891:
UserWarning: DataFrame columns are not unique, some columns will be omitted.
  "columns will be omitted.", UserWarning)
Out[6]: {'a': {0: 2}, 'b': {0: 3}}
```

#### Problem description

warning should specific stacklevel to warn at calling site, xref #9584

### PR DESCRIPTION
**Pull request by alanbato**
- [x] closes #16927 
 - [ ] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff`` (On Windows, ``git diff upstream/master -u -- "*.py" | flake8 --diff`` might work as an alternative.)
 - [x] The warning now correctly points to the calling site.

I would like some guidance on how to implement testing for this quick fix. I was thinking on using pytest's [warning assertion](https://docs.pytest.org/en/latest/warnings.html#warns), but couldn't figure out how to make it elegantly.

Cheers! :beers:

### CODE REVIEW

**Comment by gfyoung** [comment]
@alanbato : Thanks for this PR!  FYI, ignore the PEP8 bot.  We are experimenting with this.

**Comment by gfyoung** [comment]
I did a quick GitHub search for this warning, and I don't see it anywhere in our tests.  Could you add a test that triggers this warning?

**Comment by gfyoung** [comment]
> I would like some guidance on how to implement testing for this quick fix. I was thinking on using pytest's warning assertion, but couldn't figure out how to make it elegantly.

Just read this after I submitted my question :smile:  Don't try to use `pytest's` warning assertion, as we have our custom one called `assert_produces_warning` from `pandas.util.testing`.

**Comment by alanbato** [comment]
Sure! should I add it to the `pandas/core/frame/tests/frame/test_convert_to.py` file below the error assertions? Or should I create a new file?

**Comment by gfyoung** [comment]
Yes, absolutely!  That's definitely the right place to put it.

File: pandas/tests/frame/test_convert_to.py [hunk 1]
```diff
@@ -216,6 +216,12 @@ def test_to_dict_errors(self, mapping):
         with pytest.raises(TypeError):
             df.to_dict(into=mapping)
 
+    def test_to_dict_not_unique_warning(self):
+        # When converting to a dict, if a column has a non-unique name it will
+        # be dropped, throwing a warning.
+        df = DataFrame([[1, 2, 3]], columns=['a', 'a', 'b'])
+        tm.assert_produces_warning(df.to_dict())
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: needs to use with 
this is a context manager.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_convert_to.py (hunk 1 above)
Comment (alanbato) [inline pandas/tests/frame/test_convert_to.py]: Sorry, working on it right now! 😅

File: pandas/tests/frame/test_convert_to.py [hunk 2]
```diff
@@ -216,6 +216,13 @@ def test_to_dict_errors(self, mapping):
         with pytest.raises(TypeError):
             df.to_dict(into=mapping)
 
+    def test_to_dict_not_unique_warning(self):
+        # When converting to a dict, if a column has a non-unique name it will
```
Comment (alanbato) [inline pandas/tests/frame/test_convert_to.py]: Done!

**Pull request merged**